sets/maps.go

```diff
@@ -0,0 +1,137 @@
+// Copyright 2023 The Authors (see AUTHORS file)
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sets
+
+// IntersectMapKeys finds the intersection of all m keys, where intersection is
+// defined as keys that exist in all m. It always returns an allocated map, even
+// if the intersection is the empty set.
+//
+// It does not modify any of the given inputs, but also does not deep copy any
+// values. That means the returned map may have keys and values that point to
+// the same objects as in the original map.
+func IntersectMapKeys[K comparable, V any](maps ...map[K]V) map[K]V {
+	if len(maps) == 0 {
+		return make(map[K]V)
+	}
+
+	// Pre-compute the maximum possible allocation to minimize allocs. Here we
```

Comment:
This seems like it could way over-allocate depending on the actual data being input.

Wonder how it benchmarks vs for example a naieve approach or a two-pass using a bloom filter.

Reply:
I think the overhead of constructing a bloom filter would be fairly high (plus complexity). What do you mean by "naive approach"?

Reply:
Don't preallocate and allow the runtime to grow the slice for you (should be same amortized runtime, worst case is 2x the memory pressure)

Reply:
or log_2(length) the allocs

Reply:
The more I think about it, the happier I am with how it is.

If you were really looking to go above and beyond, you could copy into a smaller array at the end before returning if memory savings would be large (maybe if over 10,000 elements and less than half full), but the user could do that themselves so its not a big deal.

Reply:
Yea - I do that for the slices because it's pretty easy to trim off remaining capacity. It's less exciting for maps.